File: kotlin/src/main/kotlin/aethermesh/protocol/MeshPacket.kt
// SPDX-License-Identifier: MIT

package aethermesh.protocol

import aethermesh.AetherMeshConstants
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.util.*

/**
 * The core packet transmitted across the Aether mesh network.
 * Every piece of data — route discovery, messages, SOS broadcasts, voice,
 * streaming, DTN bundles — travels as a MeshPacket.
 *
 * Wire format is compatible with the C# implementation.
 */
data class MeshPacket(
    var id: UUID = UUID.randomUUID(),
    var type: PacketType = PacketType.Data,
    var sourceUhid: String = "",
    var destinationUhid: String = "",
    var ttl: Int = AetherMeshConstants.DEFAULT_TTL,
    var priority: Byte = 0,
    var payload: ByteArray = ByteArray(0),
    var createdAt: Long = System.currentTimeMillis(),
    var signature: ByteArray = ByteArray(0),
    var packetNonce: ByteArray = ByteArray(0),
    var timestampMs: Long = System.currentTimeMillis(),
    var protocolVersion: Byte = AetherMeshConstants.PROTOCOL_VERSION_CURRENT.toByte()
) {
    /**
     * Returns true if this packet has exceeded the maximum allowed age.
     */
    fun isExpired(maxAgeSeconds: Int = AetherMeshConstants.MAX_PACKET_AGE_SECONDS): Boolean {
        val ageMs = System.currentTimeMillis() - timestampMs
        return ageMs > maxAgeSeconds * 1000L
    }

    /**
     * Returns true if the packet can still be forwarded (TTL > 0).
     */
    fun canForward(): Boolean = ttl > 0

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is MeshPacket) return false

        if (id != other.id) return false
        if (type != other.type) return false
        if (sourceUhid != other.sourceUhid) return false
        if (destinationUhid != other.destinationUhid) return false
        if (ttl != other.ttl) return false
        if (priority != other.priority) return false
        if (!payload.contentEquals(other.payload)) return false
        if (createdAt != other.createdAt) return false
        if (!signature.contentEquals(other.signature)) return false
        if (!packetNonce.contentEquals(other.packetNonce)) return false
        if (timestampMs != other.timestampMs) return false
        if (protocolVersion != other.protocolVersion) return false

        return true
    }

    override fun hashCode(): Int {
        var result = id.hashCode()
        result = 31 * result + type.hashCode()
        result = 31 * result + sourceUhid.hashCode()
        result = 31 * result + destinationUhid.hashCode()
        result = 31 * result + ttl
        result = 31 * result + priority
        result = 31 * result + payload.contentHashCode()
        result = 31 * result + createdAt.hashCode()
        result = 31 * result + signature.contentHashCode()
        result = 31 * result + packetNonce.contentHashCode()
        result = 31 * result + timestampMs.hashCode()
        result = 31 * result + protocolVersion
        return result
    }

    override fun toString(): String {
        return "[${type.name}] ${id} src=$sourceUhid dst=$destinationUhid ttl=$ttl pri=$priority ver=$protocolVersion"
    }
}

File: swift/Sources/AetherNetProtocol/Protocol/MeshPacket.swift
// SPDX-License-Identifier: MIT

import Foundation

/// The core packet transmitted across the Aether mesh network.
/// Every piece of data — route discovery, messages, SOS broadcasts, voice,
/// streaming, DTN bundles — travels as a MeshPacket.
public struct MeshPacket: Equatable {
    /// Unique identifier for this packet.
    public var id: UUID

    /// The type of packet, determining how the payload is interpreted.
    public var type: PacketType

    /// Universal Hardware Identifier of the source node.
    public var sourceUhid: String

    /// Universal Hardware Identifier of the destination node. Empty for broadcast.
    public var destinationUhid: String

    /// Time-to-live: decremented at each hop. Packet is dropped when TTL reaches 0.
    /// Wire format is 4-byte little-endian Int32; the model field matches that width
    /// to avoid silent truncation on values > 255 (was a `UInt8` bug, fixed 2026-05-02).
    public var ttl: Int32

    /// Priority level (higher = more urgent). SOS packets use priority 255.
    public var priority: UInt8

    /// The packet payload. Interpretation depends on `type`.
    public var payload: Data

    /// UTC timestamp when this packet was created.
    public var createdAt: Date

    /// Cryptographic signature over the packet contents, produced by the source node.
    public var signature: Data

    /// Random nonce to prevent replay attacks. Must be unique per packet.
    public var packetNonce: Data

    /// Unix timestamp in milliseconds, used for age-based deduplication.
    public var timestampMs: Int64

    /// Protocol version. Current version is 2.
    public var protocolVersion: UInt8

    public init(
        id: UUID = UUID(),
        type: PacketType,
        sourceUhid: String = "",
        destinationUhid: String = "",
        ttl: Int32 = ProtocolConstants.defaultTtl,
        priority: UInt8 = 0,
        payload: Data = Data(),
        createdAt: Date = Date(),
        signature: Data = Data(),
        packetNonce: Data = Data(),
        timestampMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000),
        protocolVersion: UInt8 = ProtocolConstants.protocolVersionSigned
    ) {
        self.id = id
        self.type = type
        self.sourceUhid = sourceUhid
        self.destinationUhid = destinationUhid
        self.ttl = ttl
        self.priority = priority
        self.payload = payload
        self.createdAt = createdAt
        self.signature = signature
        self.packetNonce = packetNonce
        self.timestampMs = timestampMs
        self.protocolVersion = protocolVersion
    }

    /// Returns true if this packet has exceeded the maximum allowed age.
    public func isExpired(maxAgeSeconds: Int = ProtocolConstants.maxPacketAgeSeconds) -> Bool {
        let ageMs = Int64(Date().timeIntervalSince1970 * 1000) - timestampMs
        return ageMs > Int64(maxAgeSeconds * 1000)
    }

    /// Returns true if the packet can still be forwarded (TTL > 0).
    public var canForward: Bool {
        ttl > 0
    }

    public var description: String {
        "[\(type)] \(id.uuidString.prefix(8)) src=\(sourceUhid) dst=\(destinationUhid) ttl=\(ttl) pri=\(priority) ver=\(protocolVersion)"
    }
}
